indexer/src/referral.ts

```diff
@@ -0,0 +1,198 @@
+import {createClient} from '@supabase/supabase-js';
+
+export type Referral = {
+  fetchReferralCode: (
+    userWalletAddress: string,
+  ) => Promise<{ code: string; success: boolean; error?: any }>;
+  storeReferralCode: (
+    userWalletAddress: string,
+  ) => Promise<{ code: string; success: boolean; error?: any }>;
+  handleReferral: (
+    referredUserWalletAddress: string,
+    referralCode: string,
+  ) => Promise<{ success: boolean; error?: any; }>;
+};
+
+// 23505 is the PostgreSQL error code for unique violation
+const UNIQUE_VALIDATION_ERROR_CODE = '23505';
+
+// PGRST116 is supabase postgresql js error when single() row is requested but no row is found
+const SUPABASE_SELECT_SINGLE_ROW_ERROR_CODE = 'PGRST116';
+
+export const createReferralService = (supabaseUrl: string, supabaseKey: string) => {
+  const options = {
+    db: { schema: 'v1_cosmos' }
+  };
+  const supabase = createClient(supabaseUrl, supabaseKey, options);
+
+  async function fetchReferralCode(userWalletAddress: string): Promise<{
+    code: string;
+    success: boolean;
+    error?: any
+  }> {
+    try {
+      const { data, error } = await supabase
+        .from('user_referral_codes')
+        .select('referral_code')
+        .eq('user_wallet_address', userWalletAddress)
+        .single();
+
+      if (error) {
+        if (error.code === SUPABASE_SELECT_SINGLE_ROW_ERROR_CODE) {
+          return { code: '', success: false, error };
+        }
+
+        console.error("Error fetching user wallet address", error);
+
+        return { code: '', success: false, error };
+      }
+
+      return data ? { code: data.referral_code, success: true } : { code: '', success: false };
+    } catch (error: any) {
+      console.error("Error fetching user wallet address", error);
+      return { code: '', success: false, error };
+    }
+  }
+
+  async function storeReferralCode(userWalletAddress: string): Promise<{
```

Comment:
One issue on this function, it allows anyone to register a referral code, so I could use a script to for example make the leaderboard of wallets all my referral

Reply:
I see, working on a fix.